### ISSUE TITLE
Core Dump in model.save_weights() (ARM)

### ISSUE THREAD

**Issue by josuuribe** [tensorflow/tensorflow#56434]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.8.0

### Custom Code

Yes

### OS Platform and Distribution

ARM8 (Raspberry Pi 4) with 8GB RAM

### Mobile device

Raspberry Pi 4 Bullseye

### Python version

3.9

### Current Behaviour?

```shell
While using model.save_weights() a core dump is produced when it's used 2 times, or sometimes at first if training steps are not so much.

This behaviour can be reproduced using ModelIntervalCheckpoint() with a low interval, the first time the weights are saved fine the second time, it usually crashes (producint a core dump) or hangs up.

This has been tested on JupyterLab and also as a Python file with errors like:

terminate called after throwing an instance of 'std::system_error'
  what():  Invalid argument
Aborted (core dumped)

python3: pthread_mutex_lock.c:81: __pthread_mutex_lock: Assertion `mutex->__data.__owner == 0' failed.
Aborted (core dumped)
```

### Standalone code to reproduce the issue

```shell
# This code uses intentionally a low interval log to reproduce the error
from tensorflow.keras.optimizers import SGD, RMSprop

memory = SequentialMemory(limit=500000, window_length=WINDOW_LENGTH)
processor = AtariProcessor()
policy = LinearAnnealedPolicy(EpsGreedyQPolicy(), 
                              attr='eps',
                              value_max=1.,
                              value_min=.1, 
                              value_test=.05,
                              nb_steps=1000000)

dqn = DQNAgent(model=model,
               nb_actions=nb_actions, 
               policy=policy,
               memory=memory, 
               processor=processor,
               nb_steps_warmup=50000,
               enable_double_dqn = True,
               enable_dueling_network = True,
               dueling_type = 'avg',
               gamma=.99,
               target_model_update=10000,
               delta_clip=1.,
               train_interval=32)

dqn.compile(RMSprop(lr=.001, momentum=0.9, epsilon=1e-06, rho=0.8), metrics=['mae'])

weights_filename = 'dqn_{}_weights.h5f'.format(env_name)
checkpoint_weights_filename = 'dqn_' + env_name + '_weights_{step}.h5f'
log_filename = 'dqn_{}_log.json'.format(env_name)
callbacks = [ModelIntervalCheckpoint(checkpoint_weights_filename, interval=100)]

dqn.fit(env, callbacks=callbacks, nb_steps=150, log_interval=10000, visualize=False, verbose=2)
```

### Relevant log output

```shell
Errors

terminate called after throwing an instance of 'std::system_error'
  what():  Invalid argument
Aborted (core dumped)

python3: pthread_mutex_lock.c:81: __pthread_mutex_lock: Assertion `mutex->__data.__owner == 0' failed.
Aborted (core dumped)
```
</details>

**Comment by sushreebarsa** [comment]
@josuuribe It is not recommended to directly install TF in Raspberry pi due to the memory constraints. Could you refer this [link](https://www.tensorflow.org/lite/guide/python) to know more on this . Thank you!

**Comment by josuuribe** [comment]
@sushreebarsa thanks for your response, I have tried this it in a RPI with 8GB RAM, so I think this is not a memory constraint, in particular the model it's very simple and works perfectly except when I try to save it.

So the main problem is that seems is not possible use **model.save_weights(..)** function, this is a problem in most scenarios and I think not related to memory constraints.

**Comment by mohantym** [comment]
Hi @josuuribe ! Sorry for the late response . Could you update Threading=0  or Threading = 1 in /etc/odbcinst.ini file . Attached relevant [thread](https://stackoverflow.com/a/39488994/11530462) for reference. Thank you!

**Comment by josuuribe** [comment]
Hi @mohantym 

I have read your comment and the stackoverflow thread, I do not hace such file **/etc/odbcinst.ini** because this is a ODBC specific configuration file, not related to my problem.

**Comment by mohantym** [comment]
@josuuribe ! Sorry for the inconvenience .Could you share the steps followed on installing Tensorflow. Thank you!

**Comment by josuuribe** [comment]
@mohantym nothing special, I downloaded the binary from [Qengineering repo ](https://github.com/Qengineering/TensorFlow-Raspberry-Pi_64-bit) and then install it via pip as usual **pip install tensorflow**

**Comment by mohantym** [comment]
@gowthamkpr ! Could you please look at this issue . Thank you!

**Comment by gowthamkpr** [comment]
@josuuribe I was not able to reproduce the error. Can you please add all the necessary imports and give the complete code for me to reproduce the issue. Thanks!

**Comment by josuuribe** [comment]
Hello @gowthamkpr 

_This is my **pip list**_

Package                      Version
---------------------------- -------------------
absl-py                      1.1.0
ale-py                       0.7.5+8f3bc3b
astunparse                   1.6.3
cachetools                   5.2.0
certifi                      2022.5.18.1
charset-normalizer           2.0.12
cloudpickle                  2.1.0
cycler                       0.11.0
flatbuffers                  20181003210633
fonttools                    4.33.3
gast                         0.5.3
google-auth                  2.7.0
google-auth-oauthlib         0.4.6
google-pasta                 0.2.0
grpcio                       1.46.3
gym                          0.24.1
gym-notices                  0.0.7
h5py                         3.7.0
idna                         3.3
importlib-metadata           4.11.4
importlib-resources          5.7.1
keras                        2.8.0
Keras-Preprocessing          1.1.2
keras-rl2                    1.0.5
kiwisolver                   1.4.3
libclang                     14.0.1
Markdown                     3.3.7
matplotlib                   3.5.2
numpy                        1.22.4
oauthlib                     3.2.0
opencv-python                4.6.0.66
opt-einsum                   3.3.0
packaging                    21.3
Pillow                       9.1.1
pip                          20.3.4
pkg-resources                0.0.0
protobuf                     3.20.0
pyasn1                       0.4.8
pyasn1-modules               0.2.8
pyparsing                    3.0.9
python-dateutil              2.8.2
PyVirtualDisplay             3.0
requests                     2.28.0
requests-oauthlib            1.3.1
rsa                          4.8
setuptools                   44.1.1
six                          1.16.0
tensorboard                  2.8.0
tensorboard-data-server      0.6.1
tensorboard-plugin-wit       1.8.1
tensorflow                   2.8.0
tensorflow-io-gcs-filesystem 0.26.0
termcolor                    1.1.0
tf-estimator-nightly         2.8.0.dev2021122109
typing-extensions            4.2.0
urllib3                      1.26.9
Werkzeug                     2.1.2
wheel                        0.37.1
wrapt                        1.14.1
zipp                         3.8.0


_This is a source code example, I have simplifed it, the first **save_weights** works the second no:_

```Python
from __future__ import division

from PIL import Image
import numpy as np
import gym

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Activation, Flatten, Convolution2D, Permute
from tensorflow.keras.optimizers import Adam
import keras.backend as K

from rl.agents.dqn import DQNAgent
from rl.policy import LinearAnnealedPolicy, BoltzmannQPolicy, EpsGreedyQPolicy
from rl.memory import SequentialMemory
from rl.core import Processor
from rl.callbacks import FileLogger, ModelIntervalCheckpoint, Visualizer

import matplotlib.pyplot as plt

INPUT_SHAPE = (84, 84)
WINDOW_LENGTH = 4

env_name = 'SpaceInvaders-v0'
env = gym.make(env_name)

np.random.seed(123)
env.seed(123)
nb_actions = env.action_space.n

import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.layers import Flatten, LeakyReLU, Input, Dense, Dropout, Conv2D, Conv2DTranspose, BatchNormalization, MaxPooling2D, Activation, concatenate, Add

def create_q_model():
    INPUT_SHAPE = (84, 84)
    WINDOW_LENGTH = 4
    inputs = Input(shape=(WINDOW_LENGTH,) + INPUT_SHAPE)
    layer0 = Permute((3,2,1))(inputs)

    layer = Conv2D(32, (8,8), strides=(4,4), activation="relu")(layer0)
    layer = Conv2D(64, (4,4), strides=(2,2), activation="relu")(layer)
    layer = Conv2D(64, (3,3), strides=(1,1), activation="relu")(layer)

    x = Flatten()(layer)
    x = Dense(256)(x)
    x = Dense(nb_actions, activation = 'linear')(x)

    return keras.Model(inputs=inputs, outputs=x)

model = create_q_model()
model.summary()

import cv2

class AtariProcessor(Processor):

    def process_observation(self, observation):
        assert observation.ndim == 3  # (height, width, channel)
        img = Image.fromarray(observation)
        img = img.resize(INPUT_SHAPE).convert('L')
        processed_observation = np.array(img)
        assert processed_observation.shape == INPUT_SHAPE
        return processed_observation.astype('uint8')

    def process_state_batch(self, batch):
        processed_batch = batch.astype('float32') / 255.
        return processed_batch

    def process_reward(self, reward):
        return np.clip(reward, -1., 1.)

from rl.callbacks import Callback


import os
from rl.callbacks import Callback

from tensorflow.keras.optimizers import SGD, RMSprop

memory = SequentialMemory(limit=1000000, window_length=WINDOW_LENGTH)
processor = AtariProcessor()

exploration_steps = 150000
policy = LinearAnnealedPolicy(EpsGreedyQPolicy(),
                              attr='eps',
                              value_max=1.,
                              value_min=.05,
                              value_test=.01,
                              nb_steps=exploration_steps)

dqn = DQNAgent(model=model,
               nb_actions=nb_actions,
               policy=policy,
               memory=memory,
               processor=processor,
               nb_steps_warmup=5000,
               enable_double_dqn = True,
               enable_dueling_network = True,
               dueling_type = 'avg',
               gamma=.99,
               target_model_update=1e-03,
               delta_clip=1.,
               batch_size=32,
               train_interval=4)

learning_rate = 2.5e-3

dqn.compile(RMSprop(lr=learning_rate), metrics=['mae'])

weights_filename = 'dqn_{}_weights.h5f'.format(env_name)
checkpoint_weights_filename = 'dqn_' + env_name + '_weights_{step}.h5f'
log_filename = 'dqn_{}_log.json'.format(env_name)
callbacks = [FileLogger(log_filename, interval=10000)]

dqn.fit(env, callbacks=callbacks, nb_steps=100, log_interval=10000, visualize=False, verbose=2) # Intended low nb_steps

dqn.save_weights(weights_filename, overwrite=True)
dqn.save_weights("another", overwrite=True) #This line fails!!
```

The **nb_steps** has been intentionally reduced to 100 to reproduce the error (it's not relevant the number of steps). 
The original code has a custom Callback with several calls to **save_weights**, this example has been simplified.
As commented the first call to save_weights works, but second fails with different errors like:

```
terminate called after throwing an instance of 'std::system_error'
  what():  Invalid argument
```

```
The futex facility returned an unexpected error code.
```

**Comment by gowthamkpr** [comment]
@josuuribe I am not able to reproduce the issue on google colab. I was able to save the weights successfully multiple times. Please take a look at the gist [here](https://colab.research.google.com/gist/gowthamkpr/e5e53295d5f39ed0324af161ad9cf255/untitled379.ipynb). Thanks!

**Comment by josuuribe** [comment]
@gowthamkpr please, read the bug description, this bug is on ARM (Rpi 4)

**Comment by gowthamkpr** [comment]
@josuuribe Sorry. Got it. Thanks.

**Comment by josuuribe** [comment]
There is not stack trace, the error output is the shown in bug description and previous messages.

**Comment by gowthamkpr** [comment]
@josuuribe Got it. Thanks!

**Comment by aniruthraj** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
